tools/read_collection.cpp

```diff
@@ -4,54 +4,79 @@
 #include <range/v3/view/iota.hpp>
 
 #include "binary_collection.hpp"
+#include "io.hpp"
+#include "memory_source.hpp"
+#include "payload_vector.hpp"
 #include "util/util.hpp"
 
 using namespace pisa;
 
 int main(int argc, char** argv)
 {
     std::string collection_file;
-    std::string map_file{};
+    std::optional<std::string> map_file{};
+    std::optional<std::string> lex_file{};
     std::ptrdiff_t first, last;
 
-    CLI::App app{"read_collection - read collections."};
+    CLI::App app{"Reads binary collection to stdout."};
     app.add_option("-c,--collection", collection_file, "Collection file name")->required();
-    app.add_option("-m,--map", map_file, "String map file name");
-    app.add_option("first", first, "Element number")->required();
-    app.add_option("last", last, "Element number");
+
+    auto maptext =
+        app.add_option("--maptext", map_file, "Text file mapping each line number (ID) to a string");
+    auto maplex = app.add_option("--maplex", lex_file, "Lexicon file mapping IDs to strings");
+    maptext->excludes(maplex);
+    maplex->excludes(maptext);
+
+    app.add_option("first", first, "First element number")->required();
+    app.add_option(
+        "last",
+        last,
+        "Open-ended end of range of elements to read. "
```

Comment:
Maybe standardize as "Last element number; [First, Last] provides a range of elements to read. If Last is not defined, only one element (First) will be read" -- Or something similar?

Reply:
Noted. Actually, I think I want to change this logic. First, the only reason why it makes sense (and probably why I did it this way) to have range [first, last) instead of [first,last] is if you want to print all, and you say: 0 <len>, which is considerably more convenient for scripting than `<len> - 1`. But instead, I think what's better is to have these options:
1. `n` read `n`-th element.
2. `n m` read between `n` and `m` inclusively.
3. `n -` read from `n` to the end.

The alternative is to have users state `n n` when they want the `n`-th and `n` if they want all.

Yet another solution would be to introduce subcommands: `entry`, `range`, `all` or something like that.

Let me know what you think about any of these.

Reply:
I like `entry, range, all` if I have to be honest :)